File: Java/Tema6_Ejercicio4.java
/*
 * TEMA 6: CADENAS DE CARACTERES - EJERCICIO 4
 * ============================================
 * Escriba un programa que extraiga una subcadena de una cadena.
 */

import java.util.Scanner;

public class Tema6_Ejercicio4 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.print("Digite una cadena: ");
        String cadena = scanner.nextLine();
        
        System.out.print("Digite el índice inicial: ");
        int inicio = scanner.nextInt();
        
        System.out.print("Digite el índice final: ");
        int fin = scanner.nextInt();
        
        System.out.println("\n=== EXTRACCIÓN DE SUBCADENA ===");
        System.out.println("Cadena: '" + cadena + "'");
        System.out.println("Índice: [" + inicio + ", " + fin + "]");
        
        if (inicio >= 0 && fin <= cadena.length() && inicio <= fin) {
            String subcadena = cadena.substring(inicio, fin);
            System.out.println("Subcadena extraída: '" + subcadena + "'");
        } else {
            System.out.println("Índices inválidos.");
        }
        
        scanner.close();
    }
}

/*
 * SALIDA ESPERADA:
 * Digite una cadena: Programación
 * Digite el índice inicial: 0
 * Digite el índice final: 7
 * 
 * === EXTRACCIÓN DE SUBCADENA ===
 * Cadena: 'Programación'
 * Índice: [0, 7]
 * Subcadena extraída: 'Program'
 */

File: JS/Tema6_Ejercicio4.js
/*
 * TEMA 6: CADENAS DE CARACTERES - EJERCICIO 4
 * ============================================
 * Extraer subcadenas con substring y slice.
 * 
 * EXPLICACIÓN:
 * - substring(inicio, fin): extrae porción
 * - slice(inicio, fin): similar, pero con índices negativos
 * - substr(inicio, longitud): obsoleto pero útil
 */

const texto = "JavaScript es muy poderoso";

console.log("=== EXTRACCIÓN DE SUBCADENAS ===");
console.log(`Texto original: "${texto}"`);

// substring
console.log(`\nsubstring(0, 10): "${texto.substring(0, 10)}"`);
console.log(`substring(11, 13): "${texto.substring(11, 13)}"`);

// slice
console.log(`\nslice(0, 10): "${texto.slice(0, 10)}"`);
console.log(`slice(-9): "${texto.slice(-9)}"`);  // Últimos 9 caracteres
console.log(`slice(11, -7): "${texto.slice(11, -7)}"`);

// Extraer palabra
function extraerPalabra(texto, inicio, fin) {
    return texto.substring(inicio, fin);
}

console.log(`\nPrimer palabra: "${extraerPalabra(texto, 0, 10)}"`);
console.log(`Última palabra: "${texto.substring(texto.lastIndexOf(' ') + 1)}"`);

/*
 * SALIDA ESPERADA:
 * === EXTRACCIÓN DE SUBCADENAS ===
 * Texto original: "JavaScript es muy poderoso"
 * 
 * substring(0, 10): "JavaScript"
 * substring(11, 13): "es"
 * 
 * slice(0, 10): "JavaScript"
 * slice(-9): "poderoso"
 * slice(11, -7): "es muy"
 * 
 * Primer palabra: "JavaScript"
 * Última palabra: "poderoso"
 */
